feat(_comp_compgen_help): enhance for smartctl

chrfranke opened the pull request:
A followup to #1425 with two commits addressing independent topics.

Should now detect these properly:
```
  -n MODE[,STATUS[,STATUS2]], --nocheck=MODE[,STATUS[,STATUS2]] (ATA, SCSI)
  -s NAME[,VALUE], --set=NAME[,VALUE]
  -B [+]FILE, --drivedb=[+]FILE                                       (ATA)
```

akinomyoga commented on bash_completion:
```diff
@@ -1694,7 +1694,7 @@ _comp_compgen_help()
         [[ $_line == *([[:blank:]])-* ]] || continue
         # transform "-f FOO, --foo=FOO" to "-f , --foo=FOO" etc
-        while [[ $_line =~ ((^|[^-])-[A-Za-z0-9?][[:space:]]+)\[?[A-Z0-9]+([,_-]+[A-Z0-9]+)?(\.\.+)?\]? ]]; do
+        while [[ $_line =~ ((^|[^-])-[A-Za-z0-9?][[:space:]]+)(\[,?)?(\+\]?)?[A-Z0-9]+([,_-]+[A-Z0-9]+)?(\.\.+)?\]* ]]; do
```
Does this correctly handle `MODE[,STATUS[,STATUS2]]`? `\[` only appears at the beginning and is not contained in any repetition (like `*`, `+`, and `{n,m}`), so `\[` is allowed only at the beginning.

Also, the special rule for `[+]`  seems specific to the case of `smartctl`. Or is this pattern `[+]` common for the output of `--help`? What does it actually mean? Is it a literal `+`?

akinomyoga replied:
Thanks.

akinomyoga commented on test/t/unit/test_unit_parse_help.py:
```diff
@@ -203,6 +203,16 @@ def test_32(self, bash):
         output = assert_bash_exec(bash, "_parse_help fn", want_output=True)
         assert output.split() == "--exclude=".split()
 
+    def test_33(self, bash):
```
Thank you.

chrfranke commented:
Sorry for CI failures. Will address these in the next days.

chrfranke commented:
Force-pushed new version, now combined in one commit. This reformats (as suggested by pre-commit) and fixes the testcases.
```diff
--- a/test/t/unit/test_unit_parse_help.py
+++ b/test/t/unit/test_unit_parse_help.py
@@ -205,15 +205,19 @@ class TestUnitParseHelp:

     def test_33(self, bash):
         """Pattern found in the output of 'smartctl --help'."""
-        assert_bash_exec(bash, "fn() { echo '-f A1[,A2[,A3]] --foo=A1[,A2[,A3]]'; }")
+        assert_bash_exec(
+            bash, "fn() { echo '-f A1[,A2[,A3]] --foo=A1[,A2[,A3]]'; }"
+        )
         output = assert_bash_exec(bash, "_parse_help fn", want_output=True)
-        assert output.split() == "--foo".split()
+        assert output.split() == "--foo=".split()

     def test_34(self, bash):
         """Pattern found in the output of 'smartctl --help'."""
-        assert_bash_exec(bash, "fn() { echo '-f [+]ARG --foo=[+]ARG'; }")
+        assert_bash_exec(
+            bash, "fn() { echo '-f [+]ARG --foo=[+]ARG'; }"
+        )
         output = assert_bash_exec(bash, "_parse_help fn", want_output=True)
-        assert output.split() == "--foo".split()
+        assert output.split() == "--foo=".split()

     def test_custom_helpopt1(self, bash):
         assert_bash_exec(bash, "fn() { [[ $1 == -h ]] && echo '-option'; }")

```

chrfranke commented:
Obviously my assumption about max allowed title length (4 indentation + 74 title = 78) was wrong.  Force-pushed a new version using the original title of this PR. The code is unchanged.

chrfranke commented:
Undoing reformatting of `test_34`. Last try for this week to make pre-commit happy :-)
